cantools/database/can/cpp_source.py

```diff
@@ -0,0 +1,710 @@
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * This file was generated by cantools version {version} {date}.
+ */
+
+#include "{header}"
+
+{definitions}\
+'''
+
+SIGNAL_DECLARATION_FMT = '''\
+/**
+ * Signal {name}.
+ * Message {message_name}
+ *
+{comment}\
+ * Range: {range}
+ * Scale: {scale}
+ * Offset: {offset}
+{additional_comments}\
+ */
+class {message_name}_{name} : public Signal<{type_name}, double> {{
+public:
+    {message_name}_{name}(const uint8_t* buffer);
+
+    virtual {type_name} Raw() override;
+    virtual bool RawInRange(const {type_name}& value) const override;
+}};
+'''
+
+MESSAGE_DECLARATION_FMT = '''\
+{signal_constructors}
+
+/**
+ * Message {database_message_name}.
+ *
+{comment}\
+ */
+class {database_message_name} : public Frame {{
+public:
+    // Constructor, empty buffer
+    {database_message_name}();
+    // Constructor to maintain ownership of input buffer
+    {database_message_name}(std::unique_ptr<uint8_t[]>&& other, const size_t size);
+    // Constructor pass in buffer and buffer size, user maintains buffer ownership after object destruction
+    {database_message_name}(uint8_t* buffer, const size_t size);
+
+    ~{database_message_name}() = default;
+
+    {database_message_name}& GetMessage() {{ 
+        return *this; 
+    }}
+ 
+    // Read-only span to access underlying buffer
+    absl::Span<const uint8_t> buffer() const {{
+        return absl::Span<const uint8_t>(&_buffer[0], _buffer_size);
+    }}
+ 
+    size_t size() const {{
+        return _buffer_size;
+    }}
+
+    // Buffer to string
+    std::string to_string() const {{
+        std::ostringstream oss;
+        oss << std::hex << std::setfill('0');
+        for (size_t i = 0; i < _buffer_size; ++i) {{
+            oss << std::setw(2) << (unsigned int)_buffer[i];
+        }}
+        return oss.str();
+    }}
+  
+    // Clear buffer
+    void clear() {{
+        std::fill_n(_buffer, _buffer_size, 0u);
+    }}
+
+{signal_setters}
+
+private:
+    std::unique_ptr<uint8_t[]> _buffer_ptr;
+    uint8_t* _buffer;
+    size_t _buffer_size;
+
+public:
+{signals}
+}};
+'''
+
+MESSAGE_SIGNAL_SETTER_DECLARATION_FMT = '''\
+bool set_{name}(const double& value);
+'''
+
+MESSAGE_DEFINITION_FMT = '''\
+{signal_definitions}
+{constructor}
+{signal_setters}
+'''
+
+SIGNAL_DEFINITION_FMT = '''\
+{message_name}_{name}::{message_name}_{name}(const uint8_t* buffer)
```

Comment:
It seems like instead of inheriting from this, you should just emit the signals functions directly in the base message class and either have the signal objects created as member variables during the constructor for the class, or create and return them when the getter is called.

Reply:
Review gist and then revisit this - all signals objects inherit from Signal and are member variables in respective Message class.